File: typescript/src/cli/ui/banner.ts
/**
 * The chat's opening screen: the WEBAGENTS wordmark and a card saying which
 * agent this is, which model it runs and what it can do (2026-09-24).
 *
 * The wordmark is the Python CLI's (`python/webagents/cli/ui/splash.py`), so
 * the two SDKs open the same way. Its letters are the palette's
 * `wordmarkFace` and its shadow strokes `wordmarkShadow` (theme.ts, "Signal");
 * on a terminal that animates, a band of the agent's lime crosses it once as
 * it appears (about half a second, never again in the session).
 *
 * ONE COLUMN IN, 79 WIDE (2026-09-25). It was indented two columns, so at 80
 * columns the art ran into the last column and the full wordmark was kept for
 * 82 and up; a standard 80-column terminal got the three-line one. One column
 * of margin each side centres the 78-column art at 80 columns and never
 * writes the last column. Narrower terminals get the three-line wordmark,
 * drawn all in the letters' colour (it has no shadow strokes).
 */

import { ESC, mix, padEnd, truncate, truncateStart, visibleWidth, wrapStyled } from './ansi';
import type { Theme } from './theme';

const WORDMARK = [
  '██╗    ██╗███████╗██████╗  █████╗  ██████╗ ███████╗███╗   ██╗████████╗███████╗',
  '██║    ██║██╔════╝██╔══██╗██╔══██╗██╔════╝ ██╔════╝████╗  ██║╚══██╔══╝██╔════╝',
  '██║ █╗ ██║█████╗  ██████╔╝███████║██║  ███╗█████╗  ██╔██╗ ██║   ██║   ███████╗',
  '██║███╗██║██╔══╝  ██╔══██╗██╔══██║██║   ██║██╔══╝  ██║╚██╗██║   ██║   ╚════██║',
  '╚███╔███╔╝███████╗██████╔╝██║  ██║╚██████╔╝███████╗██║ ╚████║   ██║   ███████║',
  ' ╚══╝╚══╝ ╚══════╝╚═════╝ ╚═╝  ╚═╝ ╚═════╝ ╚══════╝╚═╝  ╚═══╝   ╚═╝   ╚══════╝',
];

const WORDMARK_SMALL = ['╦ ╦╔═╗╔╗ ╔═╗╔═╗╔═╗╔╗╔╔╦╗╔═╗', '║║║║╣ ╠╩╗╠═╣║ ╦║╣ ║║║ ║ ╚═╗', '╚╩╝╚═╝╚═╝╩ ╩╚═╝╚═╝╝╚╝ ╩ ╚═╝'];

/** The columns the full wordmark needs: one of margin, 78 of art, and the last column left empty. */
export const FULL_WORDMARK_COLUMNS = 80;
const INDENT = ' ';

function artFor(columns: number): string[] {
  return columns >= FULL_WORDMARK_COLUMNS ? WORDMARK : WORDMARK_SMALL;
}

/** The wordmark lines for this width, with the band of light at `sweep` (-1 = none). */
export function wordmark(theme: Theme, columns: number, sweep = -1): string[] {
  const art = artFor(columns);
  const { paint, palette } = theme;
  if (!paint.on) return art.map((line) => `${INDENT}${line}`);
  // The full art's letters are its `█` cells and the rest its shadow; the
  // three-line art is all letters (file comment).
  const allFace = art === WORDMARK_SMALL;
  return art.map((line, row) => {
    let out = INDENT;
    Array.from(line).forEach((ch, col) => {
      if (ch === ' ') {
        out += ' ';
        return;
      }
      const position = col + row * 2;
      let colour = allFace || ch === '█' ? palette.wordmarkFace : palette.wordmarkShadow;
      if (sweep >= 0) {
        const distance = Math.abs(position - sweep);
        if (distance < 6) colour = mix(colour, palette.agent, ((Math.cos((distance / 6) * Math.PI) + 1) / 2) * 0.85);
      }
      out += `${ESC}${paint.sgr(colour)}m${ch}`;
    });
    return `${out}${ESC}39m`;
  });
}

/** Draws the wordmark, with the light crossing it once when the terminal animates. */
export async function playWordmark(out: NodeJS.WriteStream, theme: Theme): Promise<void> {
  const columns = out.columns ?? 80;
  const lines = wordmark(theme, columns);
  out.write(`\n${lines.join('\n')}\n`);
  if (!theme.animate || !theme.paint.on) return;
  const art = artFor(columns);
  const span = Math.max(...art.map((line) => line.length)) + art.length * 2;
  const frames = 18;
  out.write(`${ESC}?25l`);
  try {
    for (let frame = 0; frame <= frames; frame += 1) {
      const sweep = frame === frames ? -1 : -6 + (frame / (frames - 1)) * (span + 12);
      out.write(`${ESC}?2026h${ESC}${lines.length}A\r${wordmark(theme, columns, sweep).join('\n')}\n${ESC}?2026l`);
      await new Promise((resolve) => setTimeout(resolve, 28));
    }
  } finally {
    out.write(`${ESC}?25h`);
  }
}

export interface WelcomeInfo {
  agent: string;
  description?: string;
  model?: string;
  tools: string[];
  folder: string;
  /** Shown in the card, in amber: a missing key, a skill that failed to load. */
  warnings: string[];
  /** The SDK's version, at the right of the card's top edge. */
  version?: string;
}

/** The card under the wordmark. */
export function welcomeCard(theme: Theme, columns: number, info: WelcomeInfo): string[] {
  const { paint, palette } = theme;
  const width = Math.max(40, Math.min(columns - 1, 80));
  const inner = width - 4;
  const border = (text: string) => paint.fg(palette.border, text);
  const title = ` ${paint.fg(palette.agent, '✦')} ${paint.bold(paint.fg(palette.text, info.agent))} `;
  const version = info.version && info.version !== 'unknown' ? ` ${paint.fg(palette.faint, `webagents ${info.version}`)} ` : '';
  const fill = Math.max(0, width - 4 - visibleWidth(title) - visibleWidth(version));
  const top = `${border('╭─')}${title}${border('─'.repeat(fill))}${version}${border('─╮')}`;
  const row = (content = '') => `${border('│')} ${padEnd(content, inner)} ${border('│')}`;
  const field = (name: string, value: string, keepEnd = false) =>
    row(`${paint.fg(palette.faint, name.padEnd(7))}${paint.fg(palette.text, (keepEnd ? truncateStart : truncate)(value, inner - 7))}`);

  const lines = [top];
  if (info.description) lines.push(row(paint.fg(palette.muted, truncate(info.description, inner))));
  lines.push(row());
  if (info.model) lines.push(field('model', info.model));
  if (info.tools.length) {
    const shown: string[] = [];
    let used = 0;
    for (const tool of info.tools) {
      if (used + tool.length + 2 > inner - 7 - 10) break;
      shown.push(tool);
      used += tool.length + 2;
    }
    const more = info.tools.length - shown.length;
    lines.push(field('tools', `${shown.join(', ')}${more > 0 ? `  +${more} more` : ''}`));
  }
  lines.push(field('folder', info.folder, true));
  for (const warning of info.warnings) {
    // Wrapped, not cut: the advice is usually the end of the sentence.
    lines.push(row());
    for (const line of wrapStyled(paint.fg(palette.warning, warning), inner, `${paint.fg(palette.warning, '▲')} `, '  ')) {
      lines.push(row(line));
    }
  }
  lines.push(border(`╰${'─'.repeat(width - 2)}╯`));
  const hint = (key: string, what: string) => `${paint.fg(palette.muted, key)} ${paint.fg(palette.faint, what)}`;
  const sep = paint.fg(palette.faint, '  ·  ');
  lines.push(
    ` ${[hint('enter', 'send'), hint('/', 'commands'), hint('↑', 'history'), hint('esc', 'stop a reply')].join(sep)}`,
  );
  return lines;
}

File: python/webagents/cli/ui/banner.py
"""
The chat's opening screen: the WEBAGENTS wordmark and a card saying which
agent this is, which model it runs and what it is made of (2026-09-24).

The wordmark is the one `splash.py` always printed, drawn the way the
TypeScript chat draws it (`typescript/src/cli/ui/banner.ts`): its letters in
the palette's `wordmark_face` and its shadow strokes in `wordmark_shadow`
(`theme.py`, "Signal"), and on a terminal that animates, a band of the agent's
lime crossing it once as it appears (about half a second, never again in the
session).

ONE COLUMN IN, 79 WIDE (2026-09-25). It was indented two columns, so at 80
columns the art ran into the last column and the full wordmark was kept for 82
and up; a standard 80-column terminal got the three-line one. One column of
margin each side centres the 78-column art at 80 columns and never writes the
last column. Narrower terminals get the three-line wordmark, all in the
letters' colour (it has no shadow strokes).
"""

from __future__ import annotations

import math
import textwrap
import time
from dataclasses import dataclass, field
from typing import List, Optional

from rich.console import Console, Group
from rich.text import Text

from .splash import WEBAGENTS_LOGO_BLOCK, WEBAGENTS_LOGO_DOUBLE
from .theme import ChatTheme, mix

_WORDMARK = WEBAGENTS_LOGO_BLOCK.strip("\n").split("\n")
_WORDMARK_SMALL = WEBAGENTS_LOGO_DOUBLE.strip("\n").split("\n")


#: The columns the full wordmark needs: one of margin, 78 of art, and the last
#: column left empty (module docstring, "One column in").
FULL_WORDMARK_COLUMNS = 80
_INDENT = " "


def _art(columns: int) -> List[str]:
    return _WORDMARK if columns >= FULL_WORDMARK_COLUMNS else _WORDMARK_SMALL


def wordmark(theme: ChatTheme, columns: int, sweep: float = -1.0) -> List[Text]:
    """The wordmark's lines; `sweep` is the band of light's position, -1 for none.

    The full art's letters are its `█` cells and the rest its shadow; the
    three-line art has no shadow strokes, so it is all letters."""
    art = _art(columns)
    p = theme.palette
    all_face = art is _WORDMARK_SMALL
    lines = []
    for row, line in enumerate(art):
        text = Text(_INDENT)
        for col, ch in enumerate(line):
            if ch == " ":
                text.append(" ")
                continue
            position = col + row * 2
            colour = p.wordmark_face if all_face or ch == "█" else p.wordmark_shadow
            if sweep >= 0:
                distance = abs(position - sweep)
                if distance < 6:
                    colour = mix(colour, p.agent, ((math.cos(distance / 6 * math.pi) + 1) / 2) * 0.85)
            text.append(ch, style=colour)
        lines.append(text)
    return lines


def play_wordmark(console: Console, theme: ChatTheme) -> None:
    """Draws the wordmark; with animation on, the light crosses it once first."""
    columns = console.width
    console.print()
    if not theme.animate or not theme.rich_colour:
        console.print(Group(*wordmark(theme, columns)))
        return
    from rich.live import Live

    art = _art(columns)
    span = max(len(line) for line in art) + len(art) * 2
    frames = 18
    with Live(Group(*wordmark(theme, columns, 0)), console=console, auto_refresh=False, transient=False) as live:
        for frame in range(frames + 1):
            sweep = -1.0 if frame == frames else -6 + frame / (frames - 1) * (span + 12)
            live.update(Group(*wordmark(theme, columns, sweep)), refresh=True)
            time.sleep(0.028)


@dataclass
class WelcomeInfo:
    agent: str
    description: str = ""
    model: str = ""
    #: The agent's tools by name, as the TypeScript card lists them.
    tools: List[str] = field(default_factory=list)
    folder: str = ""
    #: Shown in the card in amber: a missing key, a skill that failed to load.
    warnings: List[str] = field(default_factory=list)
    version: Optional[str] = None


def _truncate(text: str, width: int) -> str:
    return text if len(text) <= width else text[: max(0, width - 1)] + "…"


def _truncate_start(text: str, width: int) -> str:
    return text if len(text) <= width else "…" + text[-max(0, width - 1):]


def welcome_card(theme: ChatTheme, columns: int, info: WelcomeInfo) -> List[Text]:
    """The card under the wordmark, and the line of keys under the card."""
    p = theme.palette
    width = max(40, min(columns - 1, 80))
    inner = width - 4

    title = Text.assemble(" ", ("✦", p.agent), " ", (info.agent, f"bold {p.text}"), " ")
    version = Text(f" webagents {info.version} ", style=p.faint) if info.version and info.version != "?" else Text("")
    top = Text("╭─", style=p.border)
    top.append_text(title)
    top.append("─" * max(0, width - 4 - title.cell_len - version.cell_len), style=p.border)
    top.append_text(version)
    top.append("─╮", style=p.border)

    def row(content: Optional[Text] = None) -> Text:
        content = content or Text("")
        line = Text("│ ", style=p.border)
        line.append_text(content)
        line.append(" " * max(0, inner - content.cell_len))
        line.append(" │", style=p.border)
        return line

    def field_row(name: str, value: str, keep_end: bool = False) -> Text:
        shown = (_truncate_start if keep_end else _truncate)(value, inner - 7)
        return row(Text.assemble((name.ljust(7), p.faint), (shown, p.text)))

    lines = [top]
    if info.description:
        lines.append(row(Text(_truncate(info.description, inner), style=p.muted)))
    lines.append(row())
    if info.model:
        lines.append(field_row("model", info.model))
    if info.tools:
        shown: List[str] = []
        used = 0
        for tool in info.tools:
            if used + len(tool) + 2 > inner - 7 - 10:
                break
            shown.append(tool)
            used += len(tool) + 2
        more = len(info.tools) - len(shown)
        lines.append(field_row("tools", ", ".join(shown) + (f"  +{more} more" if more > 0 else "")))
    if info.folder:
        lines.append(field_row("folder", info.folder, keep_end=True))
    for warning in info.warnings:
        # Wrapped, not cut: the advice is usually the end of the sentence.
        lines.append(row())
        for index, part in enumerate(textwrap.wrap(warning, max(10, inner - 2)) or [""]):
            lines.append(row(Text.assemble(("▲ " if index == 0 else "  ", p.warning), (part, p.warning))))
    lines.append(Text(f"╰{'─' * (width - 2)}╯", style=p.border))

    hints = Text(" ")
    for index, (key, what) in enumerate((("enter", "send"), ("/", "commands"), ("↑", "history"), ("esc", "stop a reply"))):
        if index:
            hints.append("  ·  ", style=p.faint)
        hints.append(key, style=p.muted)
        hints.append(f" {what}", style=p.faint)
    lines.append(hints)
    return lines
